src/appleseed/renderer/modeling/camera/multilenscamera.cpp

```diff
@@ -0,0 +1,1328 @@
+            // The autofocus considers the scene at the middle of the shutter interval.
+            const float time = get_shutter_middle_time();
+            const Transformd transform = m_transform_sequence.evaluate(time);
+
+            Vector3d film_point = ndc_to_camera(m_autofocus_target);
+
+            Vector3d lens_point(0, 0, exit_pupil_center_z);
+
+            Ray3d testing_ray(film_point, normalize(lens_point - film_point));
+
+            if (!trace_ray_from_film(testing_ray))
+            {
+                RENDERER_LOG_INFO(
+                    "camera \"%s\": autofocus could not be computed. Set focal distance to infinity.",
+                    get_path().c_str());
+
+                return 1.0e38;
+            }
+
+            // Create a ray that goes through the center of the lens.
+            ShadingRay ray;
+            ray.m_org = transform.point_to_parent(testing_ray.m_org);
+            ray.m_dir = normalize(transform.vector_to_parent(testing_ray.m_dir));
+            ray.m_tmin = 0.0;
+            ray.m_tmax = std::numeric_limits<double>::max();
+            ray.m_time =
+                ShadingRay::Time::create_with_normalized_time(
+                    0.5f,
+                    get_shutter_open_begin_time(),
+                    get_shutter_close_end_time());
+            ray.m_flags = VisibilityFlags::ProbeRay;
+            ray.m_depth = 0;
+
+            // Trace the ray.
+            ShadingPoint shading_point;
+            intersector.trace(ray, shading_point);
+
+            if (shading_point.hit_surface())
+            {
+                // Hit: compute the focal distance.
+                const Vector3d v = shading_point.get_point() - ray.m_org;
+                const double af_focal_distance = -transform.vector_to_local(v).z;
+
+                RENDERER_LOG_INFO(
+                    "camera \"%s\": autofocus sets focal distance to %f (using camera position at time=%.1f).",
+                    get_path().c_str(),
+                    af_focal_distance,
+                    ray.m_time.m_absolute);
+
+                return af_focal_distance;
+            }
+            else
+            {
+                // Miss: focus at infinity.
+                RENDERER_LOG_INFO(
+                    "camera \"%s\": autofocus sets focal distance to infinity (using camera position at time=%.1f).",
+                    get_path().c_str(),
+                    ray.m_time.m_absolute);
+
+                return 1.0e38;
+            }
+        }
+        
+        Vector3d sample_exit_pupil(SamplingContext& sampling_context) const
+        {
+            sampling_context.split_in_place(2, 1);
+            const Vector2d s = sampling_context.next2<Vector2d>();
+            const Vector2d lens_point =  exit_pupil_radius * sample_disk_uniform(s);
+            return Vector3d(lens_point.x, lens_point.y, exit_pupil_center_z);
+        }
+
+        Vector3d sample_entrance_pupil(SamplingContext& sampling_context) const
```

Comment:
Are those two functions mergable?

Reply:
Yes, what is the preferred way to do so? Having an enum or boolean flag to distinguish between entrance pupil and exit pupil?